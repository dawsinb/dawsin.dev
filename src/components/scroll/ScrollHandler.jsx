import { useEffect } from 'react';
import useStore from 'Utils/store';

function ScrollHandler() {
  // get update functions from the store
  const applyScrollDelta = useStore((state) => state.applyScrollDelta);
  const snapScrollPosition = useStore((state) => state.snapScrollPosition);

  // id of current timeout so old timeouts can be cleared as scrolling happens
  let timeout = null;

  /* scrolling via mouse wheel */

  const handleWheel = (event) => {
    // adjust scroll position
    applyScrollDelta(event.deltaY / 100 / 8);
    // clear timeout if one exists
    if (timeout) {
      clearTimeout(timeout);
    }
    // call snap function after short delay
    timeout = setTimeout(snapScrollPosition, 500);
  };

  /* scrolling via touch */

  // portion of screen needed to be scrolled for a full transition
  const scrollFactor = 0.3;

  // tracker for previous touch y position so we can calculate how much
  let previousY;

  // initialize previous y on start
  const handleTouchstart = (event) => {
    // only handle first touch
    if (event.touches.length === 1) {
      previousY = event.touches.item(0).clientY;
    }
  };

  // apply normalized delta on touch move
  const handleTouchmove = (event) => {
    // calculate delta
    const scrollDelta = previousY - event.touches.item(0).clientY;

    // apply scroll normalized by the window height
    applyScrollDelta(scrollDelta / scrollFactor / window.innerHeight);

    previousY = event.touches.item(0).clientY;
  };

  // call snap function when no more touches
  const handleTouchend = (event) => {
    if (event.touches.lenghth === 0) {
      // clear timeout if one exists
      if (timeout) {
        clearTimeout(timeout);
      }
      // call snap function after short delay
      timeout = setTimeout(snapScrollPosition, 500);
    }
  };

  // register event listeners
  useEffect(() => {
    window.addEventListener('wheel', handleWheel);
    window.addEventListener('touchstart', handleTouchstart);
    window.addEventListener('touchmove', handleTouchmove);
    window.addEventListener('touchend', handleTouchend);

    return () => {
      window.removeEventListener('wheel', handleWheel);
      window.removeEventListener('touchstart', handleTouchstart);
      window.removeEventListener('touchmove', handleTouchmove);
      window.removeEventListener('touchend', handleTouchend);
    };
  }, []);

  return null;
}

export default ScrollHandler;
