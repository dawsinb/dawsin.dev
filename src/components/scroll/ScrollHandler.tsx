/**
 * @module Components/ScrollHandler
 * @mergeTarget
 */

import { useEffect } from 'react';
import { useScroll } from 'Stores/scroll';
import { ScrollOverlay } from 'Components/scroll/overlay/ScrollOverlay';

/** Props for {@link ScrollHandler} */
interface ScrollHandlerProps {
  /** Number of sections, needed to set max scroll range */
  numSections: number;
  /** Optional list of titles for each section to be displayed in {@link ScrollOverlay} */
  sectionNames?: string[];
  /** Optional list of japanese titles for each section to be displayed in {@link ScrollOverlay} */
  sectionNamesJp?: string[];
  /** Amount of the page a single movement of the scroll wheel scrolls
   *
   *  *defaults to 1 / 12*
   */
  wheelStrength?: number;
  /**
   * Amount that the movement of a touch is multiplied by.
   * In other words `1 / wheelStrength` gives the percentage of the screen that must be moved for a full page scroll
   *
   * *defaults to 3.5*
   */
  touchStrength?: number;
}
/**
 * Creates a scroll handler which updates the global scroll store. Can handle both mouse wheel/trackpad and touch scrolling.
 *
 * Also creates an interactable {@link ScrollOverlay scroll overlay} to show scroll position and allow jumping to {@link Section sections}
 *
 * The scroll store can be accessed using the {@link useTransientScroll} hook or directly using {@link useScroll}
 * @param props
 * @category Component
 */
function ScrollHandler({
  numSections,
  sectionNames = [],
  sectionNamesJp = [],
  wheelStrength = 1 / 14,
  touchStrength = 3.5
}: ScrollHandlerProps) {
  // update scroll store with number of sections
  useScroll.setState({ maxScroll: numSections - 1 });

  // get update functions from the store
  const applyScrollDelta = useScroll((state) => state.applyScrollDelta);
  const snapScrollPosition = useScroll((state) => state.snapScrollPosition);

  // id of current timeout so old timeouts can be cleared as scrolling happens
  let timeout: NodeJS.Timeout;

  /* scrolling via mouse wheel */

  const handleWheel = (event: WheelEvent) => {
    // adjust scroll position
    applyScrollDelta((event.deltaY / 100) * wheelStrength);

    // set/reset timeout
    clearTimeout(timeout);
    timeout = setTimeout(snapScrollPosition, 500);
  };

  /* scrolling via touch */

  // tracker for previous touch y position so we can calculate how much
  let previousY: number;
  let activeTouch: Touch;

  // initialize previous y on start
  const handleTouchstart = (event: TouchEvent) => {
    // update active touch on only an initial touch event
    if (event.touches.length === 1) {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      activeTouch = event.touches.item(0)!;
      previousY = activeTouch.clientY;
    }
  };

  // update scroll on move
  const handleTouchmove = (event: TouchEvent) => {
    // ensure we are tracking the active touch
    if (event.touches.item(0)?.identifier === activeTouch.identifier) {
      // calculate and apply normalized delta
      const scrollDelta = previousY - activeTouch.clientY;
      applyScrollDelta((scrollDelta / window.innerHeight) * touchStrength);

      // update previous position
      previousY = activeTouch.clientY;
    }
  };

  // call snap function when no more touches
  const handleTouchend = (event: TouchEvent) => {
    if (event.touches.length === 0) {
      // set/reset timeout
      clearTimeout(timeout);
      timeout = setTimeout(snapScrollPosition, 500);
    }
  };

  // register event listeners
  useEffect(() => {
    window.addEventListener('wheel', handleWheel);
    window.addEventListener('touchstart', handleTouchstart);
    window.addEventListener('touchmove', handleTouchmove);
    window.addEventListener('touchend', handleTouchend);
  }, []);

  return <ScrollOverlay numSections={numSections} sectionNames={sectionNames} sectionNamesJp={sectionNamesJp} />;
}

export { ScrollHandler };
export type { ScrollHandlerProps };
