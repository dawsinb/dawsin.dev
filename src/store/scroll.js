import { clamp } from 'Utils/math';

const scroll = (set, get) => ({
  scrollPosition: 0,
  minScroll: 0,
  maxScroll: 0,
  applyScrollDelta: (delta) => {
    set({
      scrollPosition: clamp(get().scrollPosition + delta, get().minScroll, get().maxScroll)
    });
  },
  snapScrollPosition: () => {
    set({
      scrollPosition: clamp(Math.round(get().scrollPosition), get().minScroll, get().maxScroll)
    });
  }
});

export default scroll;
