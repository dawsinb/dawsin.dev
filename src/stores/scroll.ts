/**
 * Global stores used to access common information across the entire app
 * @module Stores
 * @mergeTarget
 */

import { clamp } from 'utils/math';
import { create, GetState, SetState } from 'zustand';
import { subscribeWithSelector, StoreApiWithSubscribeWithSelector } from 'zustand/middleware';

/**  Variables stored in the {@link useScroll scroll store} */
interface ScrollState {
  /** Current scroll position of the page. Starts at 0 and each whole number represents a full screen scroll */
  scrollPosition: number;
  /** The lower bound of scroll position. *defaults to 0* */
  minScroll: number;
  /** The upper bound of scroll position. *defaults to 0* */
  maxScroll: number;
  /** Adds the given number to the scroll position */
  applyScrollDelta: (delta: number) => void;
  /** Snaps the scroll position to the nearest whole number */
  snapScrollPosition: () => void;
}

/**
 * Accessor function for the store used to hold and update the scroll position, alongside scrolling parameters.
 *
 * See {@link ScrollState} for the variables it contains.
 *
 * *To set up a transient subscription to the scroll position that doesn't cause rerenders use the {@link useTransientScroll} hook*
 * @category Store
 */
const useScroll = create(
  subscribeWithSelector<
    ScrollState,
    SetState<ScrollState>,
    GetState<ScrollState>,
    StoreApiWithSubscribeWithSelector<ScrollState>
  >((set, get) => ({
    scrollPosition: 0,
    minScroll: 0,
    maxScroll: 0,
    applyScrollDelta: (delta: number) => {
      set({
        scrollPosition: clamp(get().scrollPosition + delta, get().minScroll, get().maxScroll)
      });
    },
    snapScrollPosition: () => {
      set({
        scrollPosition: clamp(Math.round(get().scrollPosition), get().minScroll, get().maxScroll)
      });
    }
  }))
);

export { useScroll };
export type { ScrollState };
