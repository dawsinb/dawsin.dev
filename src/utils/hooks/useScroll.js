import { useRef, useEffect } from 'react';
import useStore from 'Utils/store';

function useScroll() {
  // store scroll position in a ref so that it is preserved on rerender and doesn't cause a rerender itself
  const scrollRef = useRef(useStore.getState().scrollPosition);

  // register transient subscription to the scroll position
  useEffect(() => {
    useStore.subscribe(
      (state) => state.scrollPosition,
      (scrollPosition) => (scrollRef.current = scrollPosition)
    );
  }, []);

  return scrollRef;
}

export default useScroll;
