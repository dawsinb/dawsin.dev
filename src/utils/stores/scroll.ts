import { clamp } from 'Utils/math';
import create, { GetState, SetState } from 'zustand';
import { subscribeWithSelector, StoreApiWithSubscribeWithSelector } from 'zustand/middleware';

type ScrollState = {
  scrollPosition: number;
  minScroll: number;
  maxScroll: number;
  applyScrollDelta: (delta: number) => void;
  snapScrollPosition: () => void;
};

const useScroll = create(
  subscribeWithSelector<
    ScrollState,
    SetState<ScrollState>,
    GetState<ScrollState>,
    StoreApiWithSubscribeWithSelector<ScrollState>
  >((set, get) => ({
    scrollPosition: 0,
    minScroll: 0,
    maxScroll: 0,
    applyScrollDelta: (delta: number) => {
      set({
        scrollPosition: clamp(get().scrollPosition + delta, get().minScroll, get().maxScroll)
      });
    },
    snapScrollPosition: () => {
      set({
        scrollPosition: clamp(Math.round(get().scrollPosition), get().minScroll, get().maxScroll)
      });
    }
  }))
);

export default useScroll;
