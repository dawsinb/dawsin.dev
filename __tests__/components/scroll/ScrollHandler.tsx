import { fireEvent, render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import useScroll from 'Utils/stores/scroll';
import {ScrollHandler} from 'Components/scroll/ScrollHandler';


describe('ScrollHandler Tests', () => {
  // define window size
  Object.defineProperty(window, 'innerWidth', { value: 1920 });
  Object.defineProperty(window, 'innerHeight', { value: 1080 });

  // scroll handler params
  const wheelStrength = 0.1;
  const touchStrength = 3.5;

  // single mouse wheel has delta of 100 positive or negative
  const deltaY = 100;

  test('Render', () => {
    const { container } = render(<ScrollHandler numSections={5} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  beforeEach(() => {
    // switch to fake timers
    jest.useFakeTimers();
    // reset scroll
    useScroll.setState({ scrollPosition: 0 });
    // render component
    render(<ScrollHandler numSections={3} wheelStrength={wheelStrength} touchStrength={touchStrength} />);
  });

  test('Wheel Scrolling', () => {
    // scrolling should start at 0
    expect(useScroll.getState().scrollPosition).toBeCloseTo(0);

    // scroll mouse wheel
    act(() => {
      // fire first wheel event
      fireEvent(
        window,
        new WheelEvent('wheel', {
          deltaY: deltaY
        })
      );
      // scroll should have happened but no snap
      expect(useScroll.getState().scrollPosition).toBeCloseTo(1 * wheelStrength);

      // fire second wheel event
      fireEvent(
        window,
        new WheelEvent('wheel', {
          deltaY: deltaY
        })
      );
      // additional scroll should have happened but no snap
      expect(useScroll.getState().scrollPosition).toBeCloseTo(2 * wheelStrength);

      // skip snap timer
      jest.runOnlyPendingTimers();
      // scroll should snap to nearest whole number
      expect(useScroll.getState().scrollPosition).toBeCloseTo(Math.round(2 * wheelStrength));
    });
  });

  test('Touch Scrolling', () => {
    // scrolling should start at 0
    expect(useScroll.getState().scrollPosition).toBeCloseTo(0);

    // create touch object
    const touch = new Touch(1, 0);

    // begin touch on screen
    act(() => {
      // fire touchstart event
      fireEvent(
        window,
        new TouchEvent('touchstart', {
          touches: new TouchList([touch])
        })
      );

      // there should be no scrolling yet
      expect(useScroll.getState().scrollPosition).toBeCloseTo(0);
    });

    // move touch up screen
    const percentageOfScreenTravelled = 1 / 8;
    act(() => {
      // update touch y
      touch.clientY = -window.innerHeight * percentageOfScreenTravelled;

      // fire touchmove event
      fireEvent(
        window,
        new TouchEvent('touchmove', {
          touches: new TouchList([touch])
        })
      );
      // there should be scrolling now
      expect(useScroll.getState().scrollPosition).toBeCloseTo(percentageOfScreenTravelled * touchStrength);
    });

    // release finger from screen
    act(() => {
      // fire touchend event
      fireEvent(
        window,
        new TouchEvent('touchend', {
          touches: new TouchList([])
        })
      );
      // there should be a delay before the snap
      expect(useScroll.getState().scrollPosition).toBeCloseTo(percentageOfScreenTravelled * touchStrength);

      // skip snap timer
      jest.runOnlyPendingTimers();
      // scroll should snap to nearest whole number
      expect(useScroll.getState().scrollPosition).toBeCloseTo(Math.round(percentageOfScreenTravelled * touchStrength));
    });
  });

  test('Out of Bounds Scrolling', () => {
    // scrolling should start at 0
    expect(useScroll.getState().scrollPosition).toBeCloseTo(0);

    // attempt to scroll below 0
    act(() => {
      // fire first wheel event
      fireEvent(
        window,
        new WheelEvent('wheel', {
          deltaY: -deltaY
        })
      );
      // scroll should be kept in bounds
      expect(useScroll.getState().scrollPosition).toBeCloseTo(0);
    });

    // set scroll to max and attempt to scroll past
    act(() => {
      // set to max scroll
      useScroll.setState({ scrollPosition: useScroll.getState().maxScroll });

      // fire first wheel event
      fireEvent(
        window,
        new WheelEvent('wheel', {
          deltaY: deltaY
        })
      );
      // scroll should be kept in bounds
      expect(useScroll.getState().scrollPosition).toBeCloseTo(useScroll.getState().maxScroll);
    });
  });
});

/* jsdom doesnt support touch events natively so we have to manually recreate them */

// create a minimal Touch class (we only care about identifier and clienty)
class Touch {
  identifier: number;
  clientY: number;

  constructor(id: number, y: number) {
    this.identifier = id;
    this.clientY = y;
  }
}

// recreate touchlist class
class TouchList {
  private touches: Touch[];
  readonly length: number;

  constructor(touches: Touch[]) {
    this.touches = touches;
    this.length = touches.length;
  }

  item(index: number): Touch | null {
    return this.touches[index];
  }
}

// recreate minimal touch event (we only need the argtype and touch list)
interface TouchEventInit extends EventInit {
  touches: TouchList;
}
class TouchEvent extends Event {
  touches: TouchList;

  constructor(typeArg: string, eventInit: TouchEventInit) {
    super(typeArg, eventInit);
    this.touches = eventInit.touches;
  }
}
