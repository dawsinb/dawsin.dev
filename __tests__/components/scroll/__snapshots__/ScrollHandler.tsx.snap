// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollHandler Tests Render 1`] = `
<div
  class="sc-gKclnd kbxgzA"
  style="--scale: 1;"
>
  <div
    class="sc-bdvvtL emXPjC"
    style="--expansion: 0; --opacity: 0;"
  />
  <div
    class="sc-eCImPb dcSRhc"
  >
    <div
      class="sc-jRQBWg bjZsJa"
      style="--shift: 0px; --rotate: 0rad;"
    >
      <svg
        viewBox="
        -1 
        -1 
        102 
        102
      "
      >
        <rect
          fill="white"
          fill-opacity="0.1"
          height="100"
          stroke="white"
          stroke-width="2"
          width="100"
        />
      </svg>
    </div>
  </div>
  <div>
    <div
      class="sc-gsDKAQ bXvaKh"
    >
      <div
        class="sc-dkPtRN dKKzpj"
      >
        <svg
          viewBox="
        -2 
        -2 
        104 
        104
      "
        >
          <circle
            cx="50"
            cy="50"
            fill="white"
            fill-opacity="0.1"
            r="50"
            stroke="white"
            stroke-width="2"
          />
        </svg>
      </div>
      <div
        class="sc-hKwDye cnVMpZ"
        style="--opacity: 0; --shift: 80px;"
      />
    </div>
    <div
      class="sc-gsDKAQ gDKndc"
    >
      <div
        class="sc-dkPtRN dKKzpj"
      >
        <svg
          viewBox="
        -1 
        -1 
        102 
        102
      "
        >
          <rect
            fill="white"
            fill-opacity="0.1"
            height="100"
            stroke="white"
            stroke-width="2"
            width="100"
          />
        </svg>
      </div>
      <div
        class="sc-hKwDye cnVMpZ"
        style="--opacity: 0; --shift: 80px;"
      />
    </div>
    <div
      class="sc-gsDKAQ ldBIMc"
    >
      <div
        class="sc-dkPtRN dKKzpj"
      >
        <svg
          viewBox="
        -2 
        -2 
        104 
        104
      "
        >
          <circle
            cx="50"
            cy="50"
            fill="white"
            fill-opacity="0.1"
            r="50"
            stroke="white"
            stroke-width="2"
          />
        </svg>
      </div>
      <div
        class="sc-hKwDye cnVMpZ"
        style="--opacity: 0; --shift: 80px;"
      />
    </div>
  </div>
</div>
`;
